enterprise/server/atime_updater/atime_updater.go

```diff
@@ -0,0 +1,236 @@
+package atime_updater
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/real_environment"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+)
+
+var (
+	// The maximum gRPC message size is 4MB. Each digest proto is
+	// 64 bytes (hash) + 8 bytes (size) + 2 bytes (tags) = 74 bytes. The
+	// remaining fields in FindMissingBlobsRequest should be small (1kb?), so
+	// we can conceivably fit 4MB / 74bytes = 14.1k of these. Cap it at 10k for
+	// now and we can increase it (or increase the update frequency) if we drop
+	// a lot of updates. Note this also affects memory usage of the proxy!
+	atimeUpdaterMaxDigestsPerUpdate = flag.Int("cache_proxy.remote_atime_max_digests_per_update", 10*1000, "The maximum number of blob digests to send in a single request for updating blob access times in the remote cache.")
+	atimeUpdaterMaxUpdatesPerGroup  = flag.Int("cache_proxy.remote_atime_max_updates_per_group", 25, "The maximum number of FindMissingBlobRequests to accumulate per group for updating blob access times in the remote cache.")
+	atimeUpdaterBatchUpdateInterval = flag.Duration("cache_proxy.remote_atime_update_interval", 30*time.Second, "The time interval to wait between sending access time updates to the remote cache.")
+	atimeUpdaterRpcTimeout          = flag.Duration("cache_proxy.remote_atime_rpc_timeout", 1*time.Second, "RPC timeout to use when updating blob access times in the remote cache.")
+)
+
+// A pending batch of atime updates (basically a FindMissingBlobsRequest).
+type atimeUpdate struct {
+	instanceName   string
+	digests        map[digest.Key]struct{} // stored as a set for de-duping.
+	digestFunction repb.DigestFunction_Value
+}
+
+func (u *atimeUpdate) toProto() *repb.FindMissingBlobsRequest {
+	req := repb.FindMissingBlobsRequest{
+		InstanceName:   u.instanceName,
+		BlobDigests:    []*repb.Digest{},
```

Comment:
you can allocate this to the correct length / capacity (to reduce allocations) since you know how long u.digests is:

either: ```
...
BlobDigests: make([]*repb.Digest, len(u.digests)),
...

for i, d := range listOfDigests {
  req.BlobDigests[i] = d.ToDigest()
}

or if you just want to do capacity:

...
BlobDigests: make([]*repb.Digest, 0, len(u.digests)),
...

for d := range u.digests {
  req.BlobDigests = append(req.BlobDigests, d.ToDigest())
} 
```

Reply:
sg, done